vercel/turborepo
# separate configs, colocate JSONSchema, remove `globalDotEnv` and `pipeline.dotEnv` from V2

### Description

This PR accomplishes a few things related to config types and schema creation:

#### Removed values from Config V2

`globalDotEnv` and `pipeline.dotEnv` are no longer valid config values.  This PR removes them.  In approaching this, I took any concession necessary to avoid changing the runtime logic (and risk breaking a migration).

Thanks to @devy-bee for bringing this to our attention in [a recent PR submission](https://github.com/vercel/turborepo/pull/8974).

#### Fully separate SchemaV1 and SchemaV2

We can still continue using all the same types (i.e. without the `V#` suffix), but now the two versions of the schema don't depend on each other.  This allows for a few useful things, but one of which is that it means we have specific types to target (e.g. `PipelineV1`, which didn't exist before) for migrations and JSON Schema validation and such.

#### JSON Schema is colocated with the types that create it

This creates the JSON Schema here, now.  This is beneficial for a lot of reasons.  I checked it into the repo (instead of making it some kind of build artifact) which means we'll have it highlighted to us when it changes (during code reviews) whereas before now it would be silent.  It also means we can use the schema locally when adding fields to experience the same validation users will experience (which was a problem before that was also solved in this repo).  It will then be trivial for turbo.build to pull out these JSON files and include them in the public assets directory.

### Testing Instructions

- `turbo.json`: `globalDotEnv` and `pipeline.dotEnv` are no longer valid  (you can use the root).
- observe the full `config-v1.ts` and `config-v2.ts` files and make sure they match your mental representation of what these configs were.
